### PR TITLE
As/async select

### PR DESCRIPTION
**Pull request by magnearun**
Adds a Async select component to the application form components

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/application/templates/parental-leave/src/forms/ParentalLeaveForm.ts [hunk 1]
```diff
@@ -24,6 +25,38 @@ import {
   getEstimatedMonthlyPay,
   getNameAndIdOfSpouse,
 } from '../fields/parentalLeaveUtils'
+import gql from 'graphql-tag'
+
+const GetUnions = gql`
```
Comment (magnearun) [inline libs/application/templates/parental-leave/src/forms/ParentalLeaveForm.ts:30]: @arnifreyrs where do we want to keep these?

File: libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const [options, setOptions] = useState<Option[]>([])
+
+  useEffect(() => {
+    async function load() {
+      const loaded = await loadOptions({ application, apolloClient })
+      setOptions(loaded)
+    }
+
+    load()
+  }, [])
```
Comment (arnifreyrs) [inline libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx:32]: you probably have to add the dependencies into the array or?

File: libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  FieldDescription,
+} from '@island.is/shared/form-fields'
+import { useLocale } from '@island.is/localization'
+import { useApolloClient } from '@apollo/client/react'
+import { Option } from '@island.is/application/core'
+
+interface Props extends FieldBaseProps {
+  field: AsyncSelectField
+}
+const AsyncSelectFormField: FC<Props> = ({ application, error, field }) => {
```
Comment (arnifreyrs) [inline libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx:19]: don't we need to know the loading state or the error from the graphql call?

File: libs/application/core/src/types/Fields.ts [hunk 4]
```diff
@@ -91,6 +99,13 @@ export interface SelectField extends BaseField {
   placeholder?: FormText
 }
 
+export interface AsyncSelectField extends BaseField {
+  readonly type: FieldTypes.ASYNC_SELECT
+  component: FieldComponents.ASYNC_SELECT
+  placeholder?: FormText
+  loadOptions: (c: Context) => Promise<any>
```
Comment (arnifreyrs) [inline libs/application/core/src/types/Fields.ts:106]: `Promise<{label: string, value: unknown}[]>`

**Review by arnifreyrs — COMMENTED** [review]
I just have some comments. I really like how we can use the client to load the options

File: libs/application/core/src/lib/fieldBuilders.ts [hunk 5]
```diff
@@ -182,6 +185,41 @@ export function buildSelectField(data: {
   }
 }
 
+export function buildAsyncSelectField(data: {
+  condition?: Condition
+  id: string
+  name: FormText
+  description?: FormText
+  placeholder?: string
+  loadOptions: (c: Context) => Promise<any>
```
[the remark this answers is not in the document]
Comment (magnearun) [reply libs/application/core/src/lib/fieldBuilders.ts:194]: yes, was going to get back to that and forgot, Promise<Option[]> right ?

File: libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx (hunk 2 above)
Comment (magnearun) [reply libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx:32]: Well...I just want it to run in the first render, but I guess the loadOptions func should be a dep ?

File: libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx (hunk 3 above)
Comment (magnearun) [reply libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx:19]: I don't think we need the loading state, options will just be a empty array until it is filled, but I'm adding the error state

File: libs/application/templates/parental-leave/src/forms/ParentalLeaveForm.ts (hunk 1 above)
Comment (magnearun) [reply libs/application/templates/parental-leave/src/forms/ParentalLeaveForm.ts:30]: yes, I'll add that

File: libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx (hunk 2 above)
Comment (arnifreyrs) [reply libs/application/ui-fields/src/lib/AsyncSelectFormField.tsx:32]: I just thought the linter would complain

**Review by arnifreyrs — APPROVED** [review]
awesome

**Pull request merged**